### ISSUE TITLE
When user types wrong date, date is not updated, but wrong text is still shown

### ISSUE THREAD

**Issue by Wrent** [elm-community/elm-datepicker#81]
When user types his own date into the datepicker, he expects it to update the date inside the datepicker too. It works fine, however if the user types invalid date (e.g. a month or a day which does not exist), it should not update the date and it should also reset the input text, so it is clear which date is selected.

It worked fine in the example Ellie app, which uses [version 7.2.3](https://ellie-app.com/5QFsDgQVva1/0), if a user types invalid date there, the input text is reset to the last known valid value.

It does not work anymore in the newest [7.2.6 version](https://ellie-app.com/p7NgZvPZba1/0), the inner date keeps unchanged, but invalid text representation is shown.

**Comment by bbqbaron** [comment]
Yes indeed. It looks like the problem could be that when submitting date text, a failure to parse passes through `Result.withDefault NoChange`, leaving the text intact.

If we were just to wipe out the input text when a submission results in `NoChange`, that would fix _this_ behavior, but FWIW I believe `SubmitText` is fired on the text input's `change`, meaning that leaving the field with a date half-written would wipe out the value. Some users might say, "Hey, damn it, I was still typing!" Does that reasoning sound right? What do you think the better experience would be?

I kind of prefer your way: if I'm typing and I leave the field, I'm _probably_ done. However, if anything ate my keyboard focus that could be kind of irritating.

**Comment by Wrent** [comment]
I totally agree that wiping the input when the user is half done could be quite irritating, however reseting it to previously valid value seems kinda fine to me (and that's how it used to work anyway).

I am not sure how often the situation "anything eats my keyboard focus" happens, but when user types some date, which is not valid, he has no way of knowing that there is a mistake and because the inner model of the datepicker still has valid (but different!) value, it could lead to submitting wrong date with the form.

So from my point of view, reseting the value to the previous one was definately safer. Maybe the datepicker could indicate, that the currectly written date is invalid, but I am not sure if it would fit to the simplicity of the library.

**Comment by bbqbaron** [comment]
Yeah, a more opinionated lib could decide to add messaging to indicate invalidity of text, but it doesn't take _that_ long to type a date, so I guess blowing it away isn't that big a cost. We can just remove it on `SubmitText` if it doesn't parse.

Paranoia: I know that the `change` event isn't treated super consistently across browsers. I can't think of a scenario in which incidental rerenders of the component could cause an erroneous `change`, but if they did in some unknown, the text field would now be super annoying to use. Just to reality-check me, have you ever seen a `change` go off without user intervention, other than something stealing focus?

If not, great! Could be a very fast bugfix (with optional added test using the existing Nightwatch tests?).

**Comment by Wrent** [comment]
> Just to reality-check me, have you ever seen a change go off without user intervention, other than something stealing focus?

No, personally I haven't experienced any unexpected `change` behavior.
